[ASCII-1513] Add json flag to diagnose command

louis-cqrl opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

This PR allow diagnose output in json format

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

https://datadoghq.atlassian.net/browse/ASCII-325

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

Run `datadog-agent diagnose -j` and `datadog-agent diagnose --json`
See if the output is a valid JSON
See if it contains the same information as `datadog-agent diagnose -v`

louis-cqrl committed:
Add JSON flag to Diagnose Config, and print it out to json

louis-cqrl committed:
Add JSON flag to Diagnose Config, and print it out to json

louis-cqrl committed:
fix linter

louis-cqrl committed:
add changelog

louis-cqrl committed:
Merge branch 'main' into louis-cqrl/json-diagnose-file

louis-cqrl committed:
fix typo

louis-cqrl commented on pkg/diagnose/connectivity/core_endpoint.go:
```diff
@@ -66,12 +66,13 @@ func Diagnose(diagCfg diagnosis.Config) []diagnosis.Diagnosis {
 
 			name := fmt.Sprintf("Connectivity to %s", url)
-			diag := createDiagnosis(name, url, "", err)
+			diag := createDiagnosis("", 200, name, url, "", err) // Hardcoded 200 status code since error is nil
```
the names seems to be the url so it is not needed

louis-cqrl commented on comp/forwarder/eventplatform/eventplatformimpl/epforwarder.go:
```diff
@@ -251,19 +252,25 @@ func Diagnose() []diagnosis.Diagnosis {
+				Diagnosis:    fmt.Sprintf("Connection to `%s` failed", url),
+				Remediation:  "Please validate Agent configuration and firewall to access " + url,
+				RawError:     err.Error(),
+				StatusCode:   400, // Hardcoded 400 status code since error is not nil
+				URL:          url,
```
they are new field of diagnosis, they are not mandatory so it should not change anything

urseberry commented on cmd/agent/subcommands/diagnose/README.md:
````diff
@@ -48,3 +49,6 @@ Normally a successful diagnosis is printed as a single dot character. If verbose
 Normally internal diagnose functions will run in the context of agent and other services. It can be overridden via –run-as-user options and if specified diagnose functions will be executed in context of the agent diagnose CLI process if possible.
+
+## ```json``` option
+Normally diagnose is displayed on stdout. If JSON option is specified, the output will be formated as JSON
````
```suggestion
Normally, diagnose is displayed on stdout. If the JSON option is specified, the output is formatted as JSON
```

urseberry commented on cmd/agent/subcommands/diagnose/README.md:
````diff
@@ -48,3 +49,6 @@ Normally a successful diagnosis is printed as a single dot character. If verbose
 
 ### ```local``` option
 Normally internal diagnose functions will run in the context of agent and other services. It can be overridden via –run-as-user options and if specified diagnose functions will be executed in context of the agent diagnose CLI process if possible.
````
```suggestion
Normally, internal diagnose functions run in the context of the Agent and other services. It can be overridden with –run-as-user options, and if specified, diagnose functions are executed in the context of the Agent diagnose CLI process if possible.
```

louis-cqrl committed:
Update cmd/agent/subcommands/diagnose/README.md

Co-authored-by: Ursula Chen <58821586+urseberry@users.noreply.github.com>

louis-cqrl committed:
Update cmd/agent/subcommands/diagnose/README.md

Co-authored-by: Ursula Chen <58821586+urseberry@users.noreply.github.com>

Kaderinho commented on cmd/agent/subcommands/diagnose/README.md:
```diff
@@ -35,6 +35,7 @@ Flags:
   -t, --list              list diagnose suites
   -l, --local             force diagnose execution by the command line instead of the agent process (useful when troubleshooting privilege related problems)
   -v, --verbose           verbose output, includes passed diagnoses, and diagnoses description
+  -j, --json              output in a json file
```
**❓ question:** you mean format here (instead of file) ?‏

louis-cqrl replied:
fixed

Kaderinho commented on cmd/agent/subcommands/diagnose/README.md:
````diff
@@ -47,4 +48,7 @@ List names of all registered diagnose suites. Output also will be filtered if in
+Normally, internal diagnose functions run in the context of the Agent and other services. It can be overridden with –run-as-user options, and if specified, diagnose functions are executed in the context of the Agent diagnose CLI process if possible.
+
+## ```json``` option
+Normally, diagnose is displayed on stdout. If the JSON option is specified, the output is formatted as JSON
````
**💭 thought:** ‏JSON is still printed on stdout right ? If yes, I don't think the first sentence is needed.

louis-cqrl replied:
yes it changed at the last minute to just display it on stdout

louis-cqrl replied:
fixed

Kaderinho commented on pkg/diagnose/runner.go:
```diff
@@ -378,6 +378,75 @@ func RunStdOut(w io.Writer, diagCfg diagnosis.Config, deps SuitesDeps) error {
+
+	diagnose, err := runDiagnose(w, diagCfg, deps)
+	if err != nil {
+		return err
```
**❓ question:** ‏are we sure that `err` is JSON formatted here ? (e.g. if I have an error when I run `--json` I might want a JSON output `{"error": "..."}`)

louis-cqrl replied:
No this not json formatted but I can do it

louis-cqrl replied:
I tried something but not sure it will work I let you check

Kaderinho reviewed (commented):
The JSON part LGTM but I'm not sure about the fields you added tbh. If technically possible I would prefer these to be hidden behind an interface (so we don't need to add new fields for new type of diagnose) 

@iglendd can we have your thoughts on this please since you designed that part 🙏

louis-cqrl commented:
> The JSON part LGTM but I'm not sure about the fields you added tbh. If technically possible I would prefer these to be hidden behind an interface (so we don't need to add new fields for new type of diagnose)
> 
> @iglendd can we have your thoughts on this please since you designed that part 🙏

I added it in the field of diagnosis cause it seems difficult to take back the informations through all different functions it goes

louis-cqrl committed:
Fix typo and update JSON error in Diagnose json flag

louis-cqrl commented on pkg/diagnose/connectivity/core_endpoint.go:
```diff
@@ -119,9 +121,12 @@ func Diagnose(diagCfg diagnosis.Config) []diagnosis.Diagnosis {
+func createDiagnosis(endpointName string, statusCode int, name string, logURL string, report string, err error) diagnosis.Diagnosis {
 	d := diagnosis.Diagnosis{
-		Name: name,
+		Name:         name,
+		StatusCode:   statusCode,
+		URL:          logURL,
+		EndpointName: endpointName,
```
Most of the time name is in this format:
"Connectivity to <endpoint url> (is Ok|failed)"
and endpointName is the name of the endpoint such as series_v2 flare etc

louis-cqrl committed:
Update releasenotes/notes/add-json-outpout-to-diagnose-command-ed472034cd227e23.yaml

Co-authored-by: maxime mouial <hush-hush@users.noreply.github.com>

louis-cqrl commented on pkg/diagnose/diagnosis/loader.go:
```diff
@@ -93,6 +94,12 @@ type Diagnosis struct {
+	// endpoint URL
+	URL string
+	// Name of the endpoint
+	EndpointName string
```
If we want to use the field name we will need to refacto a bit diagnose code, since as I talked above name field is just use to store a whole string that does not fit to what we want to display

louis-cqrl committed:
Update Diagnose json flag description in README.md

louis-cqrl committed:
Update diagnosis result codes in epforwarder.go, core_endpoint.go, metadata.go, loader.go, and runner.go

mwdd146980 committed:
add port-conflict diagnose suite (#25209)

Co-authored-by: GustavoCaso <gustavo.caso@datadoghq.com>

ogaca-dd committed:
[ASCII-1480] Remove `getDiagnose` function (#24640)

* Remove SuiteDeps from ListStdOut

* Remove diagnose.Run in api component (Use either CLI or InProcess version)

* As `RunningInAgentProcess` is false, calling `getSuitesForCLIProcess` is equivalent to calling `getSuites`

* Simplify cluster agent diagnose command
Create RunStdOutAutodiscovery that runs only check "connectivity-datadog-autodiscovery"

* Remove diagnose.SuitesDeps from GetClusterAgentDiagnose

* Fix cmdDiagnose

* Remove `RunningInAgentProcess`

* Remove getDiagnose

* Inline some functions

* Add comment to improve API component

* Replace RunStdOutAutodiscovery by RunStdOutLocalCheck

* Improve code after merge from origin/main

louis-cqrl committed:
Merge remote-tracking branch 'origin/main' into louis-cqrl/json-diagnose-file
